kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
+        self.assertFalse(orphan.user_can_read(self.own_classroom_admin))
+        self.assertFalse(orphan.user_can_read(self.own_classroom_coach))
+        self.assertFalse(orphan.user_can_read(self.member))
+
+    def test_facility_user_can_update_self(self):
+        """ A FacilityUser can update its own FacilityUser model """
+        self.assertTrue(self.member.user_can_update(self.member))
+        self.assertTrue(self.own_classroom_coach.user_can_update(self.own_classroom_coach))
+        self.assertTrue(self.own_classroom_admin.user_can_update(self.own_classroom_admin))
+        self.assertTrue(self.data["facilityadmin"].user_can_update(self.data["facilityadmin"]))
+
+    def test_admins_but_not_coaches_can_update_facility_users(self):
+        """ Users with admin (but not coach) role for a FacilityUser can update that FacilityUser """
+        self.assertTrue(self.member.user_can_update(self.data["facilityadmin"]))
+        self.assertFalse(self.member.user_can_update(self.data["facilitycoach"]))
+        self.assertTrue(self.member.user_can_update(self.own_classroom_admin))
+        self.assertFalse(self.member.user_can_update(self.own_classroom_coach))
+
+    def test_admins_and_coaches_for_other_classrooms_cannot_update_facility_users(self):
+        """ Users without admin/coach role for a specific FacilityUser cannot update that FacilityUser """
+        self.assertFalse(self.other_member.user_can_update(self.own_classroom_coach))
+        self.assertFalse(self.other_member.user_can_update(self.own_classroom_admin))
+
+    def test_only_facility_admins_can_update_unaffiliated_facility_users(self):
+        """ Only Facility admins can update FacilityUser that is not a member of a Classroom or LearnerGroup """
+        orphan = self.data["unattached_users"][0]
+        self.assertTrue(orphan.user_can_update(self.data["facilityadmin"]))
+        self.assertFalse(orphan.user_can_update(self.data["facilitycoach"]))
+        self.assertFalse(orphan.user_can_update(self.own_classroom_admin))
+        self.assertFalse(orphan.user_can_update(self.own_classroom_coach))
+        self.assertFalse(orphan.user_can_update(self.member))
+
+    def test_facility_user_cannot_delete_self(self):
+        """ A FacilityUser cannot delete its own FacilityUser model, even for an admin/coach """
+        self.assertFalse(self.member.user_can_delete(self.member))
+        self.assertFalse(self.own_classroom_coach.user_can_delete(self.own_classroom_coach))
+        self.assertFalse(self.own_classroom_admin.user_can_delete(self.own_classroom_admin))
+        self.assertFalse(self.data["facilityadmin"].user_can_delete(self.data["facilityadmin"]))
+
+    def test_only_facility_admins_can_delete_facility_user(self):
+        """ The only FacilityUsers who can delete a FacilityUser are admins for the Facility """
+        self.assertTrue(self.member.user_can_delete(self.data["facilityadmin"]))
```

Comment:
Yeah, the naming didn't feel quite right. I'll adapt it to the `user.can_delete(object)` format you suggest, as it does read more naturally. These methods can then defer to the `object.permissions.*` methods as appropriate. (I would hesitate to use the `has_perm` naming, as it could get confused with the built-in Django permissions methods).